### DISCUSSION TITLE
Usage with extension or plugin

### DISCUSSION THREAD

**Discussion by simon-milleto** [maputnik/editor#884] [Ideas]
Hello! First of all, thank you for this great tool! In my use case I have maps with multi-levels, my layers are filtered by level and I have to update manually the filter to show/hide layers based on their level, the filter look like this:
```
[
  "any",
  ["==", ["get", "level"], "3"],
  [
    "all",
    ["has", "min_level"],
    ["has", "max_level"],
    [">=", 3, ["get", "min_level"]],
    ["<=", 3, ["get", "max_level"]]
  ]
]
  ```
  If I want to show another level of this layer I have to manually update the filter, do the same when I select another layer, etc... 

Which is a lot of time consuming, I was asking myself if there is anything to do about it, a solution would be to add a level selector over the map which automatically update the filter. 

I could fork the project and do the dev on my own but another mid/long term solution would be to add the ability to add some sort of plugin or use browser extension to enhance the tool, which can also be useful if someone else have a custom use case like this.

Have you ever thought about this kind of behaviour ? I can help make this kind of feature

**Comment by HarelM** [comment]
Thanks for taking the time to write this up!
There are a few dimensions here worth discussing I believe.
When you mention plugins, the first thing that comes to my mind is maplibre-gl-js plugins such as pmtiles and contours, I believe those can be very helpful to users.
Since your use case is very specific (although I'm sure you are not the only one suffering from it) a plugin is a great idea.
But I'm not sure if it's not an overkill to develop a plugin system to allow adding plugins to maputnik or have an easy way to integrate custom code.
In any case, if you would like to push this forward it will be great, I think an initial design document we can talk about would be ideal.

**Reply by simon-milleto** [reply]
You are right, a plugin system would be a huge task and by now I have no idea on how to implement it within this app 😅 
Perhaps a good place to start would be to have a kind of global API that exposes useful methods of Maputnik app, for example expose a `updateStyle()` (which will update state `mapStyle`), a `selectLayer()` method, etc...
It could be a quick win to offer external code to interact with Maputnik. What is your opinion on this?

**Reply by HarelM** [reply]
Style is saved in local storage, you can interact with it that way if needed, might be a hack, but can work already now I guess...

**Reply by simon-milleto** [reply]
Yes that's what I saw but it's not reactive, I mean if the local storage is updated it will not be reflected directly in the app, neither in the panel or the map

**Reply by HarelM** [reply]
The desktop version has some socket connected to it to reflect file updates, there is probably a way to hack it to work for someone who really desires.
